### DISCUSSION TITLE
Go to the next hit after Find references (Multi buffer results)

### DISCUSSION THREAD

**Discussion by E-ricus** [zed-industries/zed#13132] [Feature Requests]
Hi!
First of all thanks for the effort in this editor, is a really nice one!

I'm not sure if is really hidden, but I have not been able to find a keybinding/command that allows you to go to the next item in the result of the references presented in the multi-buffer.
I think that might be the same behaviour in general for any result presented in a multi-buffer.
Right now, I can emulate it by searching for the word and going through each occurrence, but it is not fully accurate if there are other components with the same name along the results not actually being an lsp reference, and is not that nice of an experience.

I wonder if it is something plan or in the works already, or if it's already there just with a non intuitive name for me.
If it is not, I'd be happy to take a look for the effort of a PR for this, as this is probably the only recurrent flow missing for me.

**Comment by SomeoneToIgnore** [comment]
I'm afraid there's no such thing for the "find all references" case, but the project search has this thing for the search hits:
![image](https://github.com/zed-industries/zed/assets/2690773/7483594f-1377-4c94-9c91-6db230d94925)

I would imagine it's not very hard to get similar controls into the "find all references" multi buffer, but have not looked at it properly to claim this firmly.

In general, it's interesting to generalize this concept, as there's also a project diagnostics multi buffer with the same issue and, hopefully, soon there will be an "uncommitted project git changes" multi buffer with the same issue.


It's somewhat orthogonal and standalone, but next Preview will ship with an outline panel, that might ease navigation around multi buffers a bit:
![image](https://github.com/zed-industries/zed/assets/2690773/74695ab6-57ab-4eb3-b296-fd82a67b9ed9)

**Reply by E-ricus** [reply]
@SomeoneToIgnore Thanks for the quick response! 
Yes I saw that behaviour in the project search, and as I mentioned I'm currently emulating that behaviour, by searching the "word" of the references queried in the multi buffer results, and going for the results, it works okay-ish but as it is not actually going through the lsp results but doing a text/regex based search, is not as accurate.

That outline pane would indeed ease the navigation a lot! it would be amazing if there is also an exposed action that can be mapped to a key, that iterates those results without the need of changing the panel and moving through there.
If that is not present and is not planned, I could take a look after this version is shipped for the effort of adding that action 🙏

**Reply by SomeoneToIgnore** [reply]
Current issue with the outline panel is that it shows outlines only, which are neither search matches, nor references (local variables or field accesses are not visible in the outlines, but those could be references), nor diagnostics, nor git diff places — I hope to add those later, but not before the initial git multi buffer appears.

Also, it's just a first iteration of the panel, so future, possibly large, refactorings might land on top, making external PRs hard to chime in.

So I would not rush with PRs in that direction just yet, albeit it would be interesting to read somebody's proposal of how all various items listed above are supposed to be fit into the same panel.

Instead, I'd rather focus on other multi buffer's controls and API for iterating over multi buffer entries — getting a small draft PR or a proposal to read sounds nice.
If we can do that, the outline panel is already able to automatically reveal the caret position inside — so seems that we'll get all that's needed.

**Comment by ccidral** [comment]
I miss that too, some keyboard-only way to quickly navigate through the Find References results. On Neovim I use Telescope, it's pretty good, I don't think Zed should necessarily mimic Telescope but it would be great we can have e.g. a keybinding for quick reference navigation in the references buffer.

**Comment by simlal** [comment]
My sloppy workaround  when in multibuffer like findAllReferences I just redo a search with ctrl-f use your configured SelectNextMatch or SelectPrevMatch binding (linux/vscode default is enter, shift-enter). For a multibuffer like project search (ctrl-shift-f), you would need to redo the ctrl-shift-f then use a custom keybding with :
```json
{
        "context": "ProjectSearchBar",
        "bindings": {
            "f3": "search::SelectNextMatch",    // I used f3 because enter just redoes the project search
            "ctrl-f3": "search::SelectPrevMatch"
        }
```

**Reply by sircoderin** [reply]
This was just introduced with editor::MoveToStartOfNextExcerpt and editor::MoveToEndOfPreviousExcerpt in 0.177.9

**Reply by 0arm** [reply]
I think theres a better way now this is what I have:

```json
  {
    "context": "Editor && multibuffer && VimControl && !menu",
    "bindings": {
      ";": "vim::GoToNextReference",
      ",": "vim::GoToPreviousReference",
    },
  },
  ```
  
  Lets your cursor be exactly on top of the reference in each buffer.
